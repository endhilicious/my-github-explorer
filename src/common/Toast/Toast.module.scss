.toast {
  position: fixed;
  top: 20px;
  left: 50%;
  transform: translateX(-50%);
  color: #fff;
  padding: 10px 20px;
  border-radius: 4px;
  z-index: 1000;
  display: none;
  align-items: center;

  &.show {
    display: flex;
  }

  .toastMessage {
    flex: 1;
  }

  .closeButton {
    background: none;
    border: none;
    color: inherit;
    cursor: pointer;
    font-size: 1rem;
    padding: 0;
    margin-left: 10px;
  }
}
