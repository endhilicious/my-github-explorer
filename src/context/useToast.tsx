import React, { createContext, useState, useContext, useCallback, ReactNode } from 'react';

import Toast from '@/common/Toast';

interface ToastContextType {
  toastData: {
    message: string;
    visible: boolean;
    type?: string;
  };
  showToast: (message: string, type?: string, time?: number) => void;
  hideToast: () => void;
}

interface ToastProviderProps {
  children: ReactNode;
}

const ToastContext = createContext<ToastContextType | undefined>(undefined);

export const useToast = (): ToastContextType => {
  const context = useContext(ToastContext);
  if (!context) {
    throw new Error('tolong periksa kembali ToastProvider anda');
  }
  return context;
};

export const ToastProvider: React.FC<ToastProviderProps> = ({ children }) => {
  const [toastData, setToastData] = useState({
    message: '',
    visible: false,
    type: 'success'
  });

  const showToast = useCallback((message: string, type = 'success', time = 3000) => {
    setToastData({ message, visible: true, type });
    setTimeout(hideToast, time);
  }, []);

  const hideToast = useCallback(() => {
    setToastData({ message: '', visible: false, type: 'success' });
  }, []);

  return (
    <ToastContext.Provider value={{ toastData, showToast, hideToast }}>
      {children}
      <Toast />
    </ToastContext.Provider>
  );
};
