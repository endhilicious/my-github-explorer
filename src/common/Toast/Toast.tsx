import React from 'react';
import cx from 'classnames';

import { useToast } from '@/context/useToast';
import styles from './Toast.module.scss';

const Toast: React.FC = () => {
  const { toastData, hideToast } = useToast();

  const handleStyleToast = () => {
    switch (toastData.type) {
      case 'warning':
        return { background: '#c87a0d', color: '#ffffff'};
      case 'danger':
        return { background: '#dc3545', color: '#ffffff'};
      case 'success':
      default:
        return { background: '#38a368', color: '#ffffff'};
    }
  }
  return (
    <div
      className={cx(styles.toast, { [styles.show]: toastData.visible })}
      style={handleStyleToast()}
    >
      <div className={styles.toastMessage}>{toastData.message}</div>
      <button className={styles.closeButton} onClick={hideToast}>
        X
      </button>
    </div>
  );
};

export default Toast;
