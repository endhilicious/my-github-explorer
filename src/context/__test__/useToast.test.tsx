import React from 'react';
import { render, fireEvent, waitFor } from '@testing-library/react';

import Toast from '@/common/Toast';
import { ToastProvider, useToast } from '../useToast';

jest.mock('../../common/Toast', () => {
  return jest.fn(() => <div data-testid="mock-toast"></div>);
});

describe('ToastProvider', () => {
  it('renders children and provides context', () => {
    const { getByTestId } = render(
      <ToastProvider>
        <div data-testid="child-component"></div>
      </ToastProvider>
    );

    expect(getByTestId('child-component')).toBeInTheDocument();
  });

  it('shows and hides toast messages', async () => {
    const TestComponent = () => {
      const { showToast, hideToast } = useToast();

      const handleShowToast = () => {
        showToast('This is a test message', 'success', 1000);
      };

      return (
        <div>
          <button onClick={handleShowToast}>Show Toast</button>
          <button onClick={hideToast}>Hide Toast</button>
        </div>
      );
    };

    const { getByText, queryByText } = render(
      <ToastProvider>
        <TestComponent />
      </ToastProvider>
    );

    expect(queryByText('This is a test message')).toBeNull();

    fireEvent.click(getByText('Hide Toast'));

    await waitFor(() => {
      expect(queryByText('This is a test message')).toBeNull();
    });
  });
});

test('Toast component snapshot', () => {
  const { asFragment } = render(
    <ToastProvider>
      <Toast />
    </ToastProvider>
  );

  expect(asFragment()).toMatchSnapshot();
});
