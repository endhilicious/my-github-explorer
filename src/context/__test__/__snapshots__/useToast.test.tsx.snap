// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toast component snapshot 1`] = `
<DocumentFragment>
  <div
    data-testid="mock-toast"
  />
  <div
    data-testid="mock-toast"
  />
</DocumentFragment>
`;
